thruster-core-async-await/src/middleware.rs

```diff
@@ -1,12 +1,26 @@
 use std::boxed::Box;
 use futures::future::Future;
 use std::pin::Pin;
+use crate::errors::ThrusterError;
 
+#[cfg(not(feature = "thruster_error_handling"))]
+pub type MiddlewareResult<C> = C;
+#[cfg(not(feature = "thruster_error_handling"))]
 pub type MiddlewareReturnValue<T> = Pin<Box<Future<Output=T> + Send + Sync>>;
+#[cfg(not(feature = "thruster_error_handling"))]
 pub type MiddlewareNext<C> = Box<Fn(C) -> Pin<Box<Future<Output=C> + Send + Sync>> + Send + Sync>;
-
+#[cfg(not(feature = "thruster_error_handling"))]
 type MiddlewareFn<C> = fn(C, Box<((Fn(C) -> Pin<Box<Future<Output=C> + Send + Sync>>) + 'static + Send + Sync)>) -> Pin<Box<Future<Output=C> + Send + Sync>>;
 
+#[cfg(feature = "thruster_error_handling")]
+pub type MiddlewareResult<C> = Result<C, ThrusterError<C>>;
```

Comment:
Instead of `ThrusterError`, could this be another generic that implements `Context` or even a `ThrusterError` trait? Another option might be to make `ThrusterError` generic over a `cause`.